### PR TITLE
Clean up Extent logic in Geo Tree readers

### PR DESCRIPTION
**Pull request by talevy**
- min and max values of coordinates were difficult to
  track, this fixes that by introducing a new Extent
  object
- Instead of re-wrapping ByteRef into a StreamInput, a stream
  input is made once
- a new getExtent() method is introduced for use by aggregations like
  geo_bounds
- re-use bounding-box containment checks

Extent object may come in handy for actual logic in the future. if/when the time comes, it 
would make sense to up-lift the class into its own file.

### CODE REVIEW

**Review by imotov — COMMENTED** [review]
I think we are moving in the right direction here, but I would like to iterate a bit more on this PR to make sure we indeed can switch to StreamInput in a reasonable manner.

File: server/src/main/java/org/elasticsearch/common/geo/EdgeTreeReader.java [hunk 1]
```diff
@@ -20,67 +20,87 @@
...
     }
 
-    public boolean crosses(int minX, int minY, int maxX, int maxY) throws IOException {
-        ByteBufferStreamInput input = new ByteBufferStreamInput(ByteBuffer.wrap(bytesRef.bytes, bytesRef.offset, bytesRef.length));
-        int thisMinX = input.readInt();
-        int thisMinY = input.readInt();
-        int thisMaxX = input.readInt();
-        int thisMaxY = input.readInt();
+    boolean containsBottomLeft(Extent extent, boolean resetInput) throws IOException {
+        if (resetInput) {
```
[the remark this answers is not in the document]
Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/EdgeTreeReader.java]: oh, woops. you're right. never. will fix

File: server/src/main/java/org/elasticsearch/common/geo/EdgeTreeReader.java [hunk 2]
```diff
@@ -128,34 +148,47 @@ private boolean containsBottomLeft(ByteBufferStreamInput input, Edge root, int m
...
-                }
+            /* no right node if rightOffset == -1 */
+            if (root.rightOffset > 0 && extent.maxY >= root.minY && crosses(readRight(root), extent)) {
+                return true;
             }
         }
         return false;
     }
 
+    static final class Extent {
```
[the remark this answers is not in the document]
Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/EdgeTreeReader.java]: ++

File: server/src/main/java/org/elasticsearch/common/geo/GeometryTreeReader.java [hunk 3]
```diff
@@ -33,28 +34,40 @@
...
 
     public GeometryTreeReader(BytesRef bytesRef) {
-        this.bytesRef = bytesRef;
+        this.input = new ByteBufferStreamInput(ByteBuffer.wrap(bytesRef.bytes, bytesRef.offset, bytesRef.length));
     }
 
-    public boolean containedInOrCrosses(int minLon, int minLat, int maxLon, int maxLat) throws IOException {
-        ByteBufferStreamInput input = new ByteBufferStreamInput(
-            ByteBuffer.wrap(bytesRef.bytes, bytesRef.offset, bytesRef.length));
+    public Extent getExtent() throws IOException {
```
Comment (imotov) [inline server/src/main/java/org/elasticsearch/common/geo/GeometryTreeReader.java]: This method is no longer needed, right?

**Review by imotov — APPROVED** [review]
I think this might be a good stopping point for this iteration. I would still love to get rid of dependency on `ByteBufferStreamInput` eventually, but we can deal with it in future iterations.

File: server/src/main/java/org/elasticsearch/common/geo/GeometryTreeReader.java (hunk 3 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/GeometryTreeReader.java]: so, this method is to be used by aggregations like `geo_bounds` in the future, so I figured I would throw it in here while I'm at it. happy to remove and only add when used

Comment (imotov) [reply server/src/main/java/org/elasticsearch/common/geo/GeometryTreeReader.java]: In this case we should probably add a test for it and replace reading individual components with invoking `Extent(StreamInput).

Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/GeometryTreeReader.java]: ++ will follow-up

**Pull request merged**